internal/app/grpcclient/impl.go

```diff
@@ -279,30 +286,31 @@ func (i *invoker) bidiStream(
 		}
 	}
 
-	// If we received an error in any of the send logic or full-duplex reads, then exit
-	if protoErr != nil {
-		result.Error = protoErr
-		return result, nil
-	}
+	var hdr metadata.MD
+	defer func() {
+		if result != nil {
+			// Set headers and trailers from the stream
+			result.ResponseHeaders = grpcutil.ConvertMetadataToProtoHeader(hdr)
+			result.ResponseTrailers = grpcutil.ConvertMetadataToProtoHeader(stream.Trailer())
+		}
+	}()
 
 	// Sends are done, close the send side of the stream
 	if err := stream.CloseSend(); err != nil {
-		result.Error = grpcutil.ConvertGrpcToProtoError(err)
-		return result, nil
+		return nil, err
 	}
 
 	// Once the send side is closed, header metadata is ready to be read
-	hdr, err := stream.Header()
+	hdr, err = stream.Header()
 	if err != nil {
 		return nil, err
```

Comment:
We probably want to return `protoErr` here, and above on line 300, if it's not nil. Generally, `stream.Receive` is the proper way to retrieve the RPC error. I'm not certain if it's returned by `CloseSend` or `Header`.

Reply:
Just to clarify, you mean something like this (and same thing for `CloseSend`)?

```
	hdr, err = stream.Header()
	if err != nil {
		result.Error = grpcutil.ConvertGrpcToProtoError(err)
		return result, nil
	}
```

One error where I was concerned is that the return of `protoErr` is after these calls, which means something could've failed above in the send loop and got set to `protoErr` but we're returning here and losing whatever that error was (unless `CloseSend` or `Header` actually returns us that error again as you alluded to)

Reply:
No I meant more like:
```go
	hdr, err = stream.Header()
	if err != nil && protoErr == nil {
		protoErr = grpcutil.ConvertGrpcToProtoError(err)
	}
```